design/security-model.md

```diff
@@ -0,0 +1,360 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: rook-ceph-osd
+subjects:
+- kind: ServiceAccount
+  name: rook-ceph-osd
+  namespace: mycluster
+```
+
+## Proposed Changes
+
+Just as we do today the cluster admin is responsible for creating the `rook-system` namespace. I propose we have a single service account in this namespace and call it `rook-system` by default. The names used are inconsequential and can be set to something different by the cluster admin.
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-system
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: rook-system
+  namespace: rook-system
+```
+
+The `rook-system` service account is responsible for launching all pods, services, daemonsets, etc. for Rook and should have enough privilege to do and nothing more. I've not audited all the RBAC rules but a good tool to do is [here](https://github.com/liggitt/audit2rbac). For example:
+
+```
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-system
+rules:
+- apiGroups: [""]
+  resources: ["pods", "services", "configmaps"]
+  verbs: [ "get", "list", "watch", "patch", "create", "update, "delete" ]
+- apiGroups: ["extensions"]
+  resources: ["deployments", "daemonsets", "replicasets"]
+  verbs: [ "get", "list", "watch", "patch", "create", "update, "delete" ]
+- apiGroups: ["apiextensions.k8s.io"]
+  resources: ["customresourcedefinitions"]
+  verbs: [ "get", "list", "watch", "patch", "create", "update, "delete" ]
+- apiGroups: ["rook.io"]
+  resources: ["*"]
+  verbs: [ "*" ]
+---
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-system
+  namespace: rook-system
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-system
+subjects:
+- kind: ServiceAccount
+  name: rook-system
+  namespace: rook-system
+```
+
+Notably absent here are privileges to set other RBAC rules read cluster-wide secrets, etc. Because the admin created the `rook-system` namespace and service account they are free to set policies on them using PSP or namespace quotas. Similar to the ones defined [here](https://github.com/rook/rook/blob/master/Documentation/kubernetes.md#rbac-for-podsecuritypolicies)).
+
+Also note that while we use a `ClusterRole` for rook-system we only use a `RoleBinding` to grant it access to the `rook-system` namespace. It does not have cluster-wide privileges.
+
+When creating a Rook cluster the cluster admin will continue to define the namespace and cluster CRD as follows:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: mycluster
+---
+apiVersion: rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: myrookcluster
+  namespace: mycluster
+	...
+```
+
+In addition we will require that the cluster-admin define a service account and role binding as follows:
+
+```
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: rook-cluster
+  namespace: mycluster
+---
+kind: Role
```

Comment:
Does the cluster-admin need to define the Role? Are there any changes to that role the cluster-admin could possibly want to make w/out knowledge about rook internals?

Reply:
yes the cluster-admin defines all roles and bindings. I think we would document the minimal set of privs needed and the admin is responsible for creating the role correctly.

Reply:
To clarify: I meant to ask if the role couldn't be defined by rook and the admin just creates the ServiceAccount and RoleBinding. I think the advantage of being able to create a role which allows to do more is outweighed by the increased usability of not having to create a role. 

On the other side, if we just provide the YAML to create the defaults, it's not really a difference :)

Reply:
I'm leaning towards letting the admin define the role. They could decide to use a single role for multiple services beyond rook.

Reply:
I'd actually advocate to have one single SA and roles for Rook as a whole. We are moving Rook to supporting multiple backends and I wouldnt want to expose these internal backend components to the users. Let alone have the users learn more about the internals and have them create roles for each internal components.

Reply:
I'm not comfortable giving the rook pods that listen on ports access to the whole set of privy that the rook-operator service account has.

Reply:
Then for those pods we should add network policies to control ingress/egress. https://kubernetes.io/docs/concepts/services-networking/network-policies/

Reply:
that seems orthogonal. restricting network traffic does not mean that we should not follow a least privileged model for service accounts.